Extend RaceTestUtils to support N runnables

ericbottard opened the pull request:
This PR fixes the way we race() more than 2 runnables.

In doing so, this now consistently highlights a failure in `OnDiscardShouldNotLeakTest.singleOrEmpty` (left failing in this PR for now), which is good news. This can also fail consistently with way less iterations than before, so this PR drops the number of iterations to 100 as well.

simonbasle commented on reactor-test/src/main/java/reactor/test/util/RaceTestUtils.java:
```diff
@@ -99,77 +101,79 @@
-			}
+	/**
+	 * Synchronizes the execution of several {@link Runnable}s as much as possible
+	 * to test race conditions. The method blocks until both have run to completion.
```
```suggestion
	 * to test race conditions. The method blocks until all have run to completion.
```

simonbasle commented on reactor-test/src/main/java/reactor/test/util/RaceTestUtils.java:
```diff
@@ -99,77 +101,79 @@
 		}
 	}
 
+	public static void race(final Runnable r1, final Runnable r2) {
```
reintroduce the javadoc, mark as deprecated and move next to also deprecated `race(Runnable, Runnable, Scheduler)`?

ericbottard replied:
👍

simonbasle commented on reactor-test/src/main/java/reactor/test/util/RaceTestUtils.java:
```diff
@@ -99,77 +101,79 @@
-	public static void race(final Runnable r1, final Runnable r2) {
-		race(r1, r2, Schedulers.single());
+	public static void race(final Runnable... rs) {
+		race(Schedulers.parallel(), rs);
```
I don't know if using the global `parallel()` is such a good idea here. `Schedulers.single()` was already debatable I think, but `Schedulers.parallel()` has a higher chance of side effects due to the fact that this is the default implicit `Scheduler` for all core operators that require a `Scheduler`.

ericbottard replied:
`Schedulers.single()` was okay-ish since the other runnable was running on the calling thread. I tried with bounded elastic at first, but encountered some problems which may actually have been unrelated, so may give it another try. Or maybe `Schedulers.elastic()`, wdyt?

simonbasle replied:
let's aim at reintroducing the `Schedulers.boundedElastic()` here 🙏

simonbasle commented on reactor-test/src/main/java/reactor/test/util/RaceTestUtils.java:
```diff
@@ -99,77 +101,79 @@
 
 		try {
-			if (!cdl.await(5, TimeUnit.SECONDS)) {
+			if (!cdl.await(50, TimeUnit.SECONDS)) {
```
is this bump of timeout on purpose, or a leftover of debugging?

ericbottard replied:
yup, shouldn't be needed at all.

simonbasle reviewed (changes requested):
with the opening of #2616, I'd explore the last WIP commit in a separate PR. added a few comments on the rest of the PR

ericbottard commented:
> with the opening of #2616, I'd explore the last WIP commit in a separate PR. added a few comments on the rest of the PR

Yes, I'm about to force push to this PR removing the last WIP commit (and addressing comments on top). I pushed to https://github.com/reactor/reactor-core/tree/2616-fix-discardrace-monosubscriber (which currently builds on top of this) for the other fix. cc @OlegDokuka

simonbasle commented on reactor-test/src/main/java/reactor/test/util/RaceTestUtils.java:
```diff
@@ -100,56 +102,65 @@
 	 * @param r1 the first runnable
 	 * @param r2 the second runnable
-	 * @param s the {@link Scheduler} on which to execute the runnables
+	 * @deprecated use {@link #race(Scheduler, Runnable...)}
```
our deprecation policy mandates that we indicate in the deprecated comment the earliest version at which this can be removed, namely `3.6.0`

simonbasle commented on reactor-test/src/main/java/reactor/test/util/RaceTestUtils.java:
```diff
@@ -100,56 +102,65 @@
 	 * to test race conditions. The method blocks until both have run to completion.
-	 * @param r1 the first runnable
-	 * @param r2 the second runnable
+	 * Kept for binary compatibility, see the varargs variant.
```
can also be hinted at with a `@see` entry

simonbasle reviewed (changes requested):
almost good for me

simonbasle commented on reactor-test/src/main/java/reactor/test/util/RaceTestUtils.java:
```diff
@@ -116,6 +116,7 @@ public static void race(final Runnable... rs) {
 	 * Kept for binary compatibility, see the varargs variant.
 	 * @param r1 the first runnable to execute
 	 * @param r2 the second runnable to execute
+	 * @see #race(Runnable...)    
```
```suggestion
	 * @see #race(Runnable...)
```

simonbasle commented on reactor-test/src/main/java/reactor/test/util/RaceTestUtils.java:
```diff
@@ -127,7 +128,7 @@ public static void race(final Runnable r1, final Runnable r2) {
 	 * @param r1 the first runnable
 	 * @param r2 the second runnable
-	 * @deprecated use {@link #race(Scheduler, Runnable...)}
+	 * @deprecated use {@link #race(Scheduler, Runnable...)}. To be removed no sooner than 3.6.0.
```
```suggestion
	 * @deprecated Use {@link #race(Scheduler, Runnable...)}. To be removed in 3.6.0, at the earliest.
```
